### ISSUE TITLE
Allow override of oom_score_adj values

### ISSUE THREAD

**Issue by jconway** [kubernetes/kubernetes#90973]
**What would you like to be added**:

Provide a Kubernetes mechanism to allow certain pods (maybe with RBAC controls on which ones) to override the `oom_score_adj` values which are currently set based on QoS heuristics. This would allow PostgreSQL pods to actively set `oom_score_adj` to the recommended values, and would make use of Burstable QoS pods a more reasonable alternative for PostgreSQL pods.

**Why is this needed**:

# PostgreSQL, OOM Killer, and Kubernetes
FWIW I have also discussed some or all of the following with several Kubernetes community members (@mrunalp; @pmorie; @jberkus; @rhatdan) and they all recommended that I submit the issue here and label it.

/sig node

## Background
For many years[1] PostgreSQL has recommended[2] avoidance of the Linux OOM Killer by avoiding memory overcommit -- i.e. by setting `vm.overcommit_memory=2`[3].

Avoidance of memory overcommit means that when a PostgreSQL backend process requests memory, and the request cannot be met, the kernel returns an error, which PostgreSQL handles appropriately. Therefore, although the offending client then receives an error from PostgreSQL, importantly the client connection is not killed, nor are any other PostgreSQL child processes (see below).

In addition, or when that is not possible, the guidance specifies changing `oom_score_adj=-1000` for the parent "postmaster" process via the privileged startup mechanism (e.g. service script or systemd unit file), and making `oom_score_adj=0` for all child processes via two environment variables that are read during child process startup. This ensures that should the OOM killer need to reap one or more processes, the postmaster will be protected[4], and the most likely candidate to get the grim reaper's axe will be a client backend. That way the damage can be minimized.

Some of the reasons for this emphasis on OOM Killer avoidance are:

* Lost committed transactions: if the postmaster (or in HA setups the controlling Patroni processes) are killed, and replication is asynchronous (which is usually the case), transactions which have been committed on the primary database may be lost entirely when the database cluster fails over to a replica.
* Lost active connections: if a client backend process is killed, the postmaster assumes shared memory may have been corrupted, and as a result it kills all active database connections and goes into crash recovery (rolls forward through transaction logs since the last checkpoint).
* Lost inflight transactions: when client backend processes are killed, transactions which have been started but not committed will be lost entirely. At that point the client application is the only source for the inflight data.
* Down time: A PostgreSQL cluster has only a single writable primary node. If it goes down, application down time is incurred.
* Reset statistics: the crash recovery process causes collected statistics to be reset (i.e. zeroed out). This affects maintenance operations such as autovacuum and autoanalyze, which in turn will cause performance degradation, or in severe cases outages (e.g. due to out of disk space). It also affects the integrity of monitoring data collected on PostgreSQL, potentially causing lost alerts.

Undoubtedly there are others that I am neglecting here.

## Issues
There are several problems related to the OOM killer when PostgreSQL is run under Kubernetes which we have identified:

* Kubernetes requires (as I understand it) `vm.overcommit_memory=1`. This leads to promiscuous overcommit behavior and is in direct contrast with PostgreSQL best practice.
* Even worse, an OOM kill can happen even when the host node does not have any memory pressure. When the memory usage of a cgroup (pod) exceeds its memory limit, the OOM killer will reap one or more processes in the cgroup.
* oom_score_adj values are almost completely out of control of the PostgreSQL pods, preventing any attempt at following the long established best practices described above.

## Some Details
With a Guaranteed QoS pod the `oom_score_adj` values are _almost_ as desired; PostgreSQL might not be targeted in a host memory pressure scenario. But the "kill if memory limit exceeded" behavior is highly undesirable from the PostgreSQL point of view.

* `oom_score_adj=-998` -- this is good, but not the recommended special value -1000 (OOM killer disabled).
* The documented environment variables are able to successfully reset `oom_score_adj=0` for the postmaster children which is also good.

With a Burstable QoS pod, `oom_score_adj` values are set very high, and with strange semantics (smaller requested memory leads to higher `oom_score_adj`). This makes PostgreSQL a prime target if/when the host node is under memory pressure. If the host node had `vm.overcommit_memory=2`, this situation would be more tolerable because OOM kills would be unlikely. However, as noted above, Kubernetes recommends/sets `vm.overcommit_memory=1`.

* cgroup constraint OOM killer behavior does not apply -- this is good
* `oom_score_adj=1000 - 10 * (percent avail mem requested)` -- (simplified calculation - actually min 2 and max 999 applies as well) this leads to very small pod getting higher score adjust value than very large one. For example a pod requesting 1% available memory will get `oom_score_adj=990` while one requesting 50% available memory will get `oom_score_adj=500`. If the OOM killer is activated, reaping processes in the small pod is unlikely to help much. This probably leads to the wrong processes being reaped first and potentially leads to more total processes being reaped.
* According to Kubernetes source code comments with respect to the Burstable QoS pod `oom_score_adj` calculation: "this is a heuristic, it won't work if a container has many small processes". PostgreSQL, by design, has one process per client connection. There may be hundreds of small idle connections as well as many larger active connections to a typical busy database.

## Desired Behavior
* (out of scope for the Kubernetes community, but I mention it for the sake of completeness) Provide a kernel mechanism to allow equivalent behavior to `vm.overcommit_memory=2`, except acting at the cgroup level. In other words, allow a process making excess memory request within a cgroup to receive an "out of memory" error instead of using the OOM Killer to enforce the constraint. This would be the ideal solution because many users seem to want Guaranteed QoS pods, but currently the memory limit enforcement via OOM killer is a problem.
* Provide a Kubernetes mechanism to allow certain pods (maybe with RBAC controls on which ones) to override the `oom_score_adj` values which are currently set based on QoS heuristics. This would allow PostgreSQL pods to actively set `oom_score_adj` to the recommended values, and would make use of Burstable QoS pods a more reasonable alternative for PostgreSQL pods.

## Footnotes
[1] The first commit I can find with related guidance is in June of 2003: https://git.postgresql.org/gitweb/?p=postgresql.git;a=blobdiff;f=doc/src/sgml/runtime.sgml;h=02befb8480c572a8ff6fb6618fc453b268fe5b11;hp=27feffde8b620cc6c01eab7d98aa0309b01b4871;hb=8d2d92c5f0169190e4963fb541340bc8c630b02f;hpb=b4117d8b1b426b9f033a97af328ffffd0ba418d1
[2] Current guidance is here: https://www.postgresql.org/docs/12/kernel-resources.html#LINUX-MEMORY-OVERCOMMIT
[3] https://www.kernel.org/doc/Documentation/vm/overcommit-accounting
[4] https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/oom.h#L9

**Comment by jberkus** [comment]
FYI, allowing vm.overcommit_memory=2 together with cgroups memory limits has been a request for Kernel developers for at least 6 years.  cgroupsv2 was supposed to deliver this, but ultimately failed; as I understand it, they fell afoul of fundamental memory management issues in Linux.

**Comment by derekwaynecarr** [comment]
Just noting that kubelet is -999, the infra/sandbox image is typically -998.  In past we explored, pods with highest priority having the ability to influence oom_score_adj.  Thanks for the detailed write-up.

**Comment by sjenning** [comment]
Using `vm.overcommit_memory` causes overcommitment to fail at `malloc` time i.e. the point of virtual memory address space growth vs normal overcommit which is more like "`malloc`, yay, it worked!, write a byte to the `malloc`'ed region triggering an on-demand page allocation, BOOOOOM killer!"

As @jberkus pointed out, it would be lovely if we could make `malloc`s fail at the cgroup level, based on `memory.limit_in_bytes`, but it is not a thing.

Many workloads expect to be able allocate much more virtual memory than they will actually access.  Only accessed pages result in an actual page allocation of physical memory.  `mmap`ing large files is an example.  Thus setting `vm.overcommit_memory=2` is generally not a good idea and `vm.overcommit_ratio` also needs to be set properly in that case, where "properly" will vary from case to case.

The `oom_score_adj` logic
https://github.com/kubernetes/kubernetes/blob/3e2ae631b4caef5f823a5e121f860c653da32ff5/pkg/kubelet/qos/policy.go#L34-L74

It admits that the heuristic doesn't work well with containers with a bunch of small processes
https://github.com/kubernetes/kubernetes/blob/3e2ae631b4caef5f823a5e121f860c653da32ff5/pkg/kubelet/qos/policy.go#L61

Other useful references
https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt (section 10 on OOM Control)

Back in the day, we added a memcg notifier so the kubelet could be notifier immediately if there was root cgroup level memory pressure and trigger eviction hopefully before the OOM killer engaged because eviction is smarter than the OOM killer
https://github.com/kubernetes/kubernetes/pull/32577
https://github.com/kubernetes/kubernetes/blob/3e2ae631b4caef5f823a5e121f860c653da32ff5/cmd/kubelet/app/options/options.go#L383

tl;dr the only kernel-native enforcement mechanisms for cgroups with no reclaimable pages and over their `memory.limit_in_bytes` is OOM killing or freezing the process until the memory allocation can be satisfied.

**Comment by jconway** [comment]
> /remove-lifecycle stale

Oops -- thanks @jberkus

**Comment by ehashman** [comment]
@jconway hi! Removing the stale tag here is preventing this issue from being closed but isn't getting anyone's attention to look at your proposal. You will need to discuss at a SIG Node meeting and/or submit a PR to create an enhancements proposal.

Note that enhancements freeze is this Thursday, May 13, so it is unlikely this will make it into the 1.22 release.

**Comment by sgreene570** [comment]
/remove-lifecycle rotten

Hey @jconway I see that this issue has sat dormant for some time, but I was wondering if you had any advice for mitigating this issue specifically when running postgresql on kubernetes? Sounds like `export PG_OOM_ADJUST_FILE=/proc/self/oom_score_adj` is really the only advisable thing to do?

Have any recent changes in postgresql changed any of the behavior described in this write up?

**Comment by jconway** [comment]
Not much has changed with kubernetes that I am aware of. I stand by my primary recommendation, namely:
`Provide a Kubernetes mechanism to allow certain pods (maybe with RBAC controls on which ones) to override the oom_score_adj values which are currently set based on QoS heuristics. This would allow PostgreSQL pods to actively set oom_score_adj to the recommended values, and would make use of Burstable QoS pods a more reasonable alternative for PostgreSQL pods.`

That said, my understanding is that kubernetes _does_ allow running with swap enabled these days, which definitely mitigates the problem. Also kube [now supports cgroupv2](https://www.crunchydata.com/blog/kube-cluster-from-scratch-on-rocky-8), so theoretically combined with a new enough kernel (>= 5.19) the TMO approach described [here](https://engineering.fb.com/2022/06/20/data-infrastructure/transparent-memory-offloading-more-memory-at-a-fraction-of-the-cost-and-power/) could be used.

But I have moved on to a new job and don't work with kube myself much at the moment, so I have not tried anything beyond what is described in the linked blog post.
